modules/core/src/controllers/controller.ts

```diff
@@ -61,71 +73,100 @@ type ControllerProps = {
       };
   dragMode?: 'pan' | 'rotate';
   inertia?: boolean | number;
-};
-
-export default class Controller {
-  ControllerState;
-  controllerState: Record<string, any> | null = null;
-  controllerStateProps: Record<string, any> | null = null;
-  transitionManager: TransitionManager;
-  _transition;
-  _events: Record<string, any> | null = null;
-  eventManager: any = null;
-  _interactionState = {
+} & TransitionProps;
+
+export type InteractionState = {
+  inTransition?: boolean;
+  isDragging?: boolean;
+  isPanning?: boolean;
+  isRotating?: boolean;
+  isZooming?: boolean;
+}
+
+export type ViewStateChangeParameters = {
+  viewState: Record<string, any>;
+  interactionState: InteractionState;
+  oldViewState?: Record<string, any>;
+}
+
+const pinchEventWorkaround: any = {};
+
+export default class Controller<ControllerState extends IViewState<ControllerState>> {
+  ControllerState: ConstructorOf<ControllerState>;
+  protected props: ControllerProps;
```

Comment:
In luma.gl I merge the props with default values then I can just rely on the props and not have to copy things over.
```suggestion
  protected props: Required<ControllerProps>;
```